Jenkinsfile

```diff
@@ -84,45 +85,48 @@ String gcpLabel(String s) {
+def runIntegrationTest(String description, String kubeprodArgs, String ginkgoArgs, Closure clusterSetup, Closure dnsSetup, Closure clusterDestroy) {
+    timeout(240) {
         // Regex of tests that are temporarily skipped.  Empty-string
         // to run everything.  Include pointers to tracking issues.
         def skip = ''
 
-        withEnv(["KUBECONFIG=${env.WORKSPACE}/.kubeconf"]) {
+        withEnv([
+            "KUBECONFIG=${env.WORKSPACE}/.kubeconf",
+            "PATH+KTOOL=${tool 'kubectl'}"
+        ]) {
             clusterSetup()
 
-            withEnv(["PATH+KTOOL=${tool 'kubectl'}"]) {
-                sh "kubectl version; kubectl cluster-info"
+            sh "kubectl version; kubectl cluster-info"
 
-                unstash 'binary'
-                unstash 'manifests'
-                unstash 'tests'
+            unstash 'binary'
+            unstash 'manifests'
+            unstash 'tests'
 
-                sh "kubectl --namespace kube-system get po,deploy,svc,ing"
+            sh "kubectl --namespace kube-system get po,deploy,svc,ing"
 
-                sh "./bin/kubeprod -v=1 install ${kubeprodArgs} --manifests=manifests"
+            sh "./bin/kubeprod -v=1 install ${kubeprodArgs} --manifests=manifests"
 
-                dnsSetup()
+            dnsSetup()
 
-                waitForRollout("kubeprod", 30)
+            waitForRollout("kubeprod", 30)
 
-                sh 'go get github.com/onsi/ginkgo/ginkgo'
-                dir('tests') {
-                    try {
-                        ansiColor('xterm') {
-                            sh "ginkgo -v --tags integration -r --randomizeAllSpecs --randomizeSuites --failOnPending --trace --progress --slowSpecThreshold=300 --compilers=2 --nodes=8 --skip '${skip}' -- --junit junit --description '${description}' --kubeconfig ${KUBECONFIG} ${ginkgoArgs}"
-                        }
-                    } catch (error) {
-                        sh "kubectl --namespace kubeprod get po,deploy,svc,ing"
-                        input 'Paused for manual debugging'
-                            throw error
-                    } finally {
-                        junit 'junit/*.xml'
+            sh 'go get github.com/onsi/ginkgo/ginkgo'
+            dir('tests') {
+                try {
+                    ansiColor('xterm') {
+                        sh "ginkgo -v --tags integration -r --randomizeAllSpecs --randomizeSuites --failOnPending --trace --progress --slowSpecThreshold=300 --compilers=2 --nodes=8 --skip '${skip}' -- --junit junit --description '${description}' --kubeconfig ${KUBECONFIG} ${ginkgoArgs}"
                     }
+                } catch (error) {
+                    sh "kubectl --namespace kubeprod get po,deploy,svc,ing"
+                    input 'Paused for manual debugging'
+                    throw error
+                } finally {
+                    junit 'junit/*.xml'
                 }
             }
+
+            clusterDestroy()
```

Comment:
With this change, `clusterDestroy()` will not be run when any of the previous steps fail (raise an exception).  I think we need to run `clusterDestroy()` in some sort of try/finally block (see the previous code structure).

Reply:
Wrapped inside a finally {}.